File: python/examples/ex05_runs.py
"""Example 05 -- task runs end-to-end: create (v3/v4) -> poll or stream SSE -> resume -> bill.

Purpose
    Task runs are server-driven: Coasty operates one of YOUR machines
    (``machine_id``) toward a natural-language ``task``, stepping the agent
    for you. This example creates a run (with an ``Idempotency-Key``), then
    follows it to a terminal state two ways:

    - default: poll ``GET /v1/runs/{id}`` until the status is terminal;
    - ``--events``: consume the durable SSE stream
      ``GET /v1/runs/{id}/events`` via :mod:`coasty.sse`, which reconnects
      automatically with ``Last-Event-ID`` (the seq cursor) so a dropped
      connection never loses or duplicates events.

    Either way, when the run pauses in ``awaiting_human`` (a captcha, a 2FA
    prompt...), you are prompted -- or ``--auto-resume NOTE`` answers
    unattended -- and the example POSTs ``/v1/runs/{id}/resume``. Billing
    events are printed as they stream, and the final summary includes the
    authoritative ``run.cost_cents``.

v3 vs v4 (the ``--v4`` flag)
    ``cua_version=v3`` is the default engine. ``--v4`` switches to the
    autonomous v4 engine with a verifier pass -- NOTE: v4 requires the pro+
    tier; other tiers get ``400 FEATURE_NOT_AVAILABLE``. Both bill the same
    5 credits per run step (only the legacy v1 engine costs 8 cr/step).

Endpoints used
    POST /v1/runs, GET /v1/runs/{id}, GET /v1/runs/{id}/events (SSE),
    POST /v1/runs/{id}/resume (scopes ``runs:read``, ``runs:write``)

Estimated cost
    5 credits ($0.05) per completed run step on v3/v4 (8 cr on v1); run steps
    carry NO surcharges. Worst case = ``--max-steps`` x the per-step rate,
    COMPUTED via :mod:`coasty.cost` and printed before any spend. Charges are
    debited up front and auto-refunded on failure; sandbox keys print
    "$0 (sandbox)". The machine's own runtime bills separately (see ex08).

Run it
    python examples/ex05_runs.py --machine-id mch_test_abc --task "Export the report" --confirm
    python examples/ex05_runs.py --machine-id mch_test_abc --task "..." --events --auto-resume ok
"""

from __future__ import annotations

import argparse
import json
import sys
import time
import uuid
from collections.abc import Callable, Sequence
from dataclasses import dataclass, field
from typing import Any, cast

from coasty import CoastyClient, CoastyError, env
from coasty.cost import CostEstimate, estimate_run
from coasty.types import TERMINAL_RUN_STATUSES, CuaVersion, Run
from ex01_local_predict_loop import spend_gate

ResumeHandler = Callable[[str | None], str | None]
"""Given the ``awaiting_human`` reason, return an optional resume note."""


@dataclass(frozen=True)
class StreamOutcome:
    """What the SSE stream delivered, plus the authoritative final run."""

    final_run: Run
    events_seen: list[str] = field(default_factory=list)  # event types, in order
    billing_events: list[dict[str, Any]] = field(default_factory=list)
    resumed: bool = False


def create_task_run(
    client: CoastyClient,
    *,
    machine_id: str,
    task: str,
    cua_version: CuaVersion = "v3",
    max_steps: int | None = None,
    idempotency_key: str | None = None,
    emit: Callable[[str], None] = print,
) -> Run:
    """POST /v1/runs with an Idempotency-Key (safe to retry, never duplicated)."""
    result = client.create_run(
        machine_id,
        task,
        cua_version=cua_version,
        max_steps=max_steps,
        # pause (the default) is what makes awaiting_human -> /resume possible;
        # "fail"/"cancel" would end the run instead of waiting for a human.
        on_awaiting_human="pause",
        idempotency_key=idempotency_key,
    )
    run = result.data
    replay = " (idempotent replay)" if result.idempotent_replay else ""
    emit(
        f"run {run['id']} created{replay}: status={run['status']} engine={run['cua_version']} "
        f"max_steps={run['max_steps']} (request_id={result.request_id})"
    )
    return run


def poll_run(
    client: CoastyClient,
    run_id: str,
    *,
    on_awaiting_human: ResumeHandler,
    poll_interval: float = 2.0,
    sleep: Callable[[float], None] = time.sleep,
    max_polls: int = 10_000,
    emit: Callable[[str], None] = print,
) -> Run:
    """Poll GET /v1/runs/{id} until terminal, resuming through awaiting_human.

    ``sleep`` is injected so tests poll instantly. Raises ``TimeoutError``
    if the run is still not terminal after ``max_polls`` polls.
    """
    for _ in range(max_polls):
        run = client.get_run(run_id).data
        status = run["status"]
        if status in TERMINAL_RUN_STATUSES:
            return run
        emit(f"run {run_id}: status={status} steps_completed={run['steps_completed']}")
        if status == "awaiting_human":
            note = on_awaiting_human(run["awaiting_human_reason"])
            # Only valid from awaiting_human; anything else would be a 409
            # NOT_AWAITING_HUMAN (surfaced as a typed ConflictError).
            client.resume_run(run_id, note=note)
            emit(f"resumed run {run_id}" + (f" with note {note!r}" if note else ""))
        sleep(poll_interval)
    raise TimeoutError(f"run {run_id} was not terminal after {max_polls} polls")


def _event_payload(data: str) -> dict[str, Any]:
    """Tolerantly decode an SSE data payload into a dict (or {})."""
    try:
        payload = json.loads(data)
    except json.JSONDecodeError:
        return {}
    return payload if isinstance(payload, dict) else {}


def stream_run_events(
    client: CoastyClient,
    run_id: str,
    *,
    on_awaiting_human: ResumeHandler,
    last_event_id: int | str | None = None,
    emit: Callable[[str], None] = print,
) -> StreamOutcome:
    """Consume GET /v1/runs/{id}/events to ``done``, resuming when paused.

    ``client.run_events`` (built on :mod:`coasty.sse`) transparently
    reconnects after a drop, sending the last seen seq as ``Last-Event-ID``
    -- events are durable server-side, so nothing is lost or duplicated.
    """
    events_seen: list[str] = []
    billing_events: list[dict[str, Any]] = []
    resumed = False

    for event in client.run_events(run_id, last_event_id=last_event_id):
        events_seen.append(event.event)
        payload = _event_payload(event.data)
        if event.event == "billing":
            billing_events.append(payload)
            emit(f"[billing seq={event.id}] {event.data}")
        elif event.event == "awaiting_human":
            reason = payload.get("reason")
            emit(f"[awaiting_human seq={event.id}] reason={reason!r}")
            note = on_awaiting_human(reason if isinstance(reason, str) else None)
            client.resume_run(run_id, note=note)
            resumed = True
            emit(f"resumed run {run_id}" + (f" with note {note!r}" if note else ""))
        elif event.event == "error":
            emit(f"[error seq={event.id}] {event.data}")
        else:
            emit(f"[{event.event} seq={event.id}] {event.data}")
        # the iterator stops by itself right after the "done" event

    final_run = client.get_run(run_id).data  # authoritative result + billing
    return StreamOutcome(
        final_run=final_run,
        events_seen=events_seen,
        billing_events=billing_events,
        resumed=resumed,
    )


def summarize_run(run: Run, *, emit: Callable[[str], None] = print) -> None:
    """Print the terminal state, result/error, and the authoritative cost."""
    emit(f"final status: {run['status']} ({run['steps_completed']} step(s) completed)")
    result = run["result"]
    if result is not None:
        emit(f"  result: passed={result['passed']} summary={result['summary']!r}")
    error = run["error"]
    if error is not None:
        emit(f"  error: {error['code']}: {error['message']}")
    emit(
        f"  billed: {run['credits_charged']} credit(s) = {run['cost_cents']} cents "
        f"(${run['cost_cents'] / 100:.2f}) [run.cost_cents is authoritative]"
    )


def build_estimate(max_steps: int, cua_version: CuaVersion) -> CostEstimate:
    """Worst case: every allowed step completes (run steps have no surcharges)."""
    return estimate_run(steps=max_steps, cua_version=cua_version)


def make_resume_handler(auto_note: str | None) -> ResumeHandler:
    """Interactive prompt by default; ``--auto-resume NOTE`` answers unattended."""

    def handler(reason: str | None) -> str | None:
        described = reason or "no reason given"
        if auto_note is not None:
            print(f"awaiting_human ({described}): auto-resuming with note {auto_note!r}")
            return auto_note
        print(f"run is awaiting_human: {described}")
        answer = input("resolve it on the machine, then press Enter (optional note): ").strip()
        return answer or None

    return handler


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Create a task run, follow it (poll or --events SSE) to a terminal state."
    )
    parser.add_argument("--machine-id", required=True, help="machine to run the task on")
    parser.add_argument("--task", required=True, help="natural-language task (1-16000 chars)")
    parser.add_argument("--max-steps", type=int, default=20, help="server-side step cap (1-1000)")
    parser.add_argument(
        "--v4",
        action="store_true",
        help="use the autonomous v4 engine (verifier pass; REQUIRES the pro+ tier -- "
        "other tiers get 400 FEATURE_NOT_AVAILABLE; still 5 cr/step)",
    )
    parser.add_argument(
        "--events", action="store_true", help="stream SSE events instead of polling"
    )
    parser.add_argument(
        "--auto-resume",
        metavar="NOTE",
        help="on awaiting_human, resume immediately with this note instead of prompting",
    )
    parser.add_argument(
        "--idempotency-key", help="Idempotency-Key for the create (default: generated)"
    )
    parser.add_argument("--poll-interval", type=float, default=2.0, help="seconds between polls")
    parser.add_argument(
        "--confirm", action="store_true", help="allow spending on a live (non-sandbox) key"
    )
    args = parser.parse_args(list(argv) if argv is not None else None)
    cua_version: CuaVersion = "v4" if args.v4 else "v3"

    try:
        api_key = env.require_api_key()
    except env.MissingAPIKeyError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    estimate = build_estimate(args.max_steps, cua_version)
    if not spend_gate(
        estimate, api_key=api_key, confirm=args.confirm, title=f"ex05 task run ({cua_version})"
    ):
        return 2
    if args.v4:
        print("note: v4 is the autonomous engine with a verifier pass (pro+ tier only).")

    idempotency_key = (
        args.idempotency_key if args.idempotency_key is not None else f"ex05-run-{uuid.uuid4().hex}"
    )
    on_awaiting_human = make_resume_handler(cast("str | None", args.auto_resume))

    try:
        with CoastyClient(api_key) as client:
            run = create_task_run(
                client,
                machine_id=args.machine_id,
                task=args.task,
                cua_version=cua_version,
                max_steps=args.max_steps,
                idempotency_key=idempotency_key,
            )
            if args.events:
                outcome = stream_run_events(client, run["id"], on_awaiting_human=on_awaiting_human)
                final_run = outcome.final_run
            else:
                final_run = poll_run(
                    client,
                    run["id"],
                    on_awaiting_human=on_awaiting_human,
                    poll_interval=args.poll_interval,
                )
    except CoastyError as exc:
        print(f"error: {exc}", file=sys.stderr)  # str() includes the request_id
        if exc.request_id:
            print(f"request_id: {exc.request_id}", file=sys.stderr)
        return 1
    except TimeoutError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1

    summarize_run(final_run)
    return 0 if final_run["status"] == "succeeded" else 1


if __name__ == "__main__":
    sys.exit(main())

File: typescript/src/examples/ex05-runs.ts
/**
 * Example 05 — Task runs: create (v3 / v4), poll OR stream, resume, billing.
 *
 * Purpose: hand a whole task to a Coasty-managed agent on one of YOUR
 * machines and supervise it. Demonstrates:
 *
 *   - POST /v1/runs with an Idempotency-Key (retried creates can never
 *     double-start a run; replays carry X-Coasty-Idempotent-Replay: true);
 *   - cua_version v3 (default) and v4 behind `--v4` — NOTE: v4 (autonomous
 *     mode + verifier) requires the pro+ tier; other tiers get
 *     400 FEATURE_NOT_AVAILABLE;
 *   - default mode: GET /v1/runs/{id} polling until a terminal status
 *     (succeeded | failed | cancelled | timed_out);
 *   - `--events` mode: GET /v1/runs/{id}/events SSE consumption with
 *     AUTOMATIC Last-Event-ID reconnect (src/coasty/sse.ts replays after the
 *     last seen seq — no loss, no duplicates);
 *   - the awaiting_human -> POST /v1/runs/{id}/resume flow in both modes;
 *   - billing events from the stream + the final result with cost_cents.
 *
 * Endpoints: POST /v1/runs, GET /v1/runs/{id}, GET /v1/runs/{id}/events,
 * POST /v1/runs/{id}/resume.
 *
 * Estimated cost (computed at runtime via src/coasty/cost.ts and printed by
 * the spend gate): estimateRunCredits({steps: maxSteps, cuaVersion}) —
 * 5 cr/step on v3/v4, 8 cr/step on v1; run steps have NO other surcharges.
 * Defaults: 10 x 5 = 50 cr = $0.50 ceiling (only completed steps bill).
 * Sandbox keys never bill — the gate prints "$0 (sandbox)".
 *
 * Run it:
 *   npx tsx src/examples/ex05-runs.ts --machine mch_test_123 \
 *     --task "Open the dashboard and export the report" --confirm
 *   npx tsx src/examples/ex05-runs.ts --machine mch_test_123 --events --v4 --confirm
 *
 * Env config: COASTY_API_KEY (required), COASTY_BASE_URL, COASTY_MACHINE_ID,
 * EX05_TASK, EX05_MAX_STEPS, COASTY_CONFIRM_SPEND=1 (instead of --confirm).
 */
import { randomUUID } from 'node:crypto';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { CoastyClient } from '../coasty/client.js';
import {
  estimateRunCredits,
  formatEstimate,
  formatUsd,
  runStepCredits,
  type EstimateLineItem,
} from '../coasty/cost.js';
import { getApiKey, isSandboxKey, spendConfirmed } from '../coasty/env.js';
import { CoastyError } from '../coasty/errors.js';
import { defaultSleep, type SleepFn } from '../coasty/sse.js';
import {
  isTerminalRunStatus,
  type CreateRunRequest,
  type CuaVersion,
  type ResumeRunRequest,
  type Run,
  type RunEvent,
} from '../coasty/types.js';

export type PrintFn = (line: string) => void;

export const stdoutPrint: PrintFn = (line) => {
  process.stdout.write(`${line}\n`);
};

export const V4_TIER_NOTE =
  'Note: cua_version v4 (autonomous mode + verifier) requires the pro+ tier — ' +
  'other tiers get 400 FEATURE_NOT_AVAILABLE. v3 is the default for all tiers.';

// ---------------------------------------------------------------------------
// Spend gate
// ---------------------------------------------------------------------------

export interface SpendGateOptions {
  apiKey: string;
  items: readonly EstimateLineItem[];
  /** True when the user passed --confirm. */
  confirmFlag: boolean;
  env?: NodeJS.ProcessEnv;
  print?: PrintFn;
}

/**
 * Print the itemized estimate, then decide whether billable calls may
 * proceed: sandbox keys always may ("$0 (sandbox)"); live keys require
 * `--confirm` or `COASTY_CONFIRM_SPEND=1`.
 */
export function ensureSpendApproved(options: SpendGateOptions): boolean {
  const print = options.print ?? stdoutPrint;
  const sandbox = isSandboxKey(options.apiKey);
  print(formatEstimate(options.items, { sandbox }));
  if (sandbox) {
    print('Spend gate: $0 (sandbox) — sandbox keys never bill; proceeding.');
    return true;
  }
  if (options.confirmFlag || spendConfirmed(options.env ?? process.env)) {
    print('Spend gate: confirmed — proceeding with billable calls.');
    return true;
  }
  print(
    'Spend gate: BLOCKED — this key bills real money. Re-run with --confirm ' +
      'or set COASTY_CONFIRM_SPEND=1 to proceed.',
  );
  return false;
}

// ---------------------------------------------------------------------------
// Create
// ---------------------------------------------------------------------------

export interface BuildRunRequestOptions {
  machineId: string;
  task: string;
  /** 'v3' (default, all tiers) or 'v4' (pro+ only — see V4_TIER_NOTE). */
  cuaVersion?: CuaVersion;
  maxSteps?: number;
}

export function buildCreateRunRequest(options: BuildRunRequestOptions): CreateRunRequest {
  return {
    machine_id: options.machineId,
    task: options.task,
    cua_version: options.cuaVersion ?? 'v3',
    max_steps: options.maxSteps ?? 10,
    // pause (default) is what makes the awaiting_human -> resume flow below
    // possible; 'fail' / 'cancel' end the run instead.
    on_awaiting_human: 'pause',
  };
}

export interface CreateRunOptions {
  client: CoastyClient;
  request: CreateRunRequest;
  /** Default: a fresh `ex05-run-<uuid>` key. */
  idempotencyKey?: string;
  print?: PrintFn;
}

/** POST /v1/runs with an Idempotency-Key; surfaces idempotent replays. */
export async function createRun(options: CreateRunOptions): Promise<Run> {
  const print = options.print ?? stdoutPrint;
  const idempotencyKey = options.idempotencyKey ?? `ex05-run-${randomUUID()}`;
  const { data, meta } = await options.client.runs.create(options.request, { idempotencyKey });
  print(
    `created run ${data.id} (status=${data.status}, cua_version=${data.cua_version}, ` +
      `max_steps=${String(data.max_steps)}, request_id=${data.request_id ?? meta.requestId ?? 'n/a'})`,
  );
  if (meta.idempotentReplay) {
    print('note: served from the idempotency cache (X-Coasty-Idempotent-Replay: true).');
  }
  return data;
}

// ---------------------------------------------------------------------------
// Poll mode
// ---------------------------------------------------------------------------

/**
 * Decide what to do when a run pauses for a human. Return a resume body to
 * POST /runs/{id}/resume, or null to keep waiting (e.g. a real operator is
 * handling it out-of-band).
 */
export type AwaitingHumanHandler = (run: Run) => Promise<ResumeRunRequest | null>;

export interface PollRunOptions {
  client: CoastyClient;
  runId: string;
  /** Delay between GETs (default 2000 ms). */
  intervalMs?: number;
  /** Injectable for tests (default: real setTimeout sleep). */
  sleep?: SleepFn;
  onAwaitingHuman?: AwaitingHumanHandler;
  /** Called on every status TRANSITION (not every poll). */
  onUpdate?: (run: Run) => void;
  /** Safety valve so a stuck run cannot poll forever (default 10000). */
  maxPolls?: number;
  print?: PrintFn;
}

/** GET /v1/runs/{id} until a terminal status; resumes awaiting_human runs. */
export async function pollRunUntilTerminal(options: PollRunOptions): Promise<Run> {
  const intervalMs = options.intervalMs ?? 2000;
  const sleep = options.sleep ?? defaultSleep;
  const maxPolls = options.maxPolls ?? 10_000;
  const print = options.print ?? stdoutPrint;

  let lastStatus: string | null = null;
  for (let poll = 1; poll <= maxPolls; poll += 1) {
    const { data: run } = await options.client.runs.get(options.runId);
    if (run.status !== lastStatus) {
      lastStatus = run.status;
      print(`run ${run.id}: ${run.status} (steps_completed=${String(run.steps_completed)})`);
      options.onUpdate?.(run);
    }
    if (isTerminalRunStatus(run.status)) return run;
    if (run.status === 'awaiting_human' && options.onAwaitingHuman !== undefined) {
      print(`awaiting_human: ${run.awaiting_human_reason ?? 'no reason given'}`);
      const resume = await options.onAwaitingHuman(run);
      if (resume !== null) {
        const { data: resumed } = await options.client.runs.resume(options.runId, resume);
        print(`resumed run ${resumed.id} (status=${resumed.status}).`);
        lastStatus = resumed.status;
      }
    }
    await sleep(intervalMs);
  }
  throw new Error(
    `run ${options.runId} did not reach a terminal status after ${String(maxPolls)} poll(s)`,
  );
}

// ---------------------------------------------------------------------------
// --events mode (SSE with automatic Last-Event-ID reconnect)
// ---------------------------------------------------------------------------

function extractReason(data: unknown): string | null {
  if (typeof data === 'object' && data !== null && 'reason' in data) {
    const reason = (data as Record<string, unknown>).reason;
    return typeof reason === 'string' ? reason : null;
  }
  return null;
}

export interface WatchRunOptions {
  client: CoastyClient;
  runId: string;
  /** Resume cursor for a previous partial watch (sent as Last-Event-ID). */
  lastEventId?: string | number;
  /** Max automatic reconnects after stream drops (default 5). */
  maxReconnects?: number;
  onAwaitingHuman?: AwaitingHumanHandler;
  /** Observe every event (tests use this; printing already happens). */
  onEvent?: (event: RunEvent) => void;
  print?: PrintFn;
}

export interface WatchRunResult {
  /** Final run object (GET after the stream's `done`). */
  run: Run;
  /** Every seq observed, in order — no loss, no duplicates across reconnects. */
  seqs: number[];
  /** Payloads of `billing` events, in order. */
  billingEvents: unknown[];
}

/**
 * Consume GET /v1/runs/{id}/events until the `done` event. The underlying
 * client stream (src/coasty/sse.ts) reconnects automatically on drops,
 * carrying `Last-Event-ID: <last seen seq>` so the durable event log is
 * replayed exactly-once. awaiting_human events trigger the resume flow
 * inline; billing events are collected and printed.
 */
export async function watchRunEvents(options: WatchRunOptions): Promise<WatchRunResult> {
  const print = options.print ?? stdoutPrint;
  const seqs: number[] = [];
  const billingEvents: unknown[] = [];

  const events = options.client.runs.events(options.runId, {
    lastEventId: options.lastEventId,
    maxReconnects: options.maxReconnects,
  });
  for await (const event of events) {
    if (event.seq >= 0) seqs.push(event.seq);
    options.onEvent?.(event);
    switch (event.type) {
      case 'billing':
        billingEvents.push(event.data);
        print(`[seq ${String(event.seq)}] billing ${JSON.stringify(event.data)}`);
        break;
      case 'awaiting_human': {
        const reason = extractReason(event.data);
        print(`[seq ${String(event.seq)}] awaiting_human: ${reason ?? 'no reason given'}`);
        if (options.onAwaitingHuman !== undefined) {
          // The stream stays open while we resume; subsequent events
          // (`resumed`, more steps, `done`) arrive on the same stream.
          const { data: current } = await options.client.runs.get(options.runId);
          const resume = await options.onAwaitingHuman(current);
          if (resume !== null) {
            const { data: resumed } = await options.client.runs.resume(options.runId, resume);
            print(`resumed run ${resumed.id} (status=${resumed.status}).`);
          }
        }
        break;
      }
      case 'error':
        print(`[seq ${String(event.seq)}] error ${JSON.stringify(event.data)}`);
        break;
      case 'done':
        print(`[seq ${String(event.seq)}] done — stream complete.`);
        break;
      default:
        print(`[seq ${String(event.seq)}] ${event.type} ${JSON.stringify(event.data)}`);
        break;
    }
  }

  const { data: run } = await options.client.runs.get(options.runId);
  return { run, seqs, billingEvents };
}

// ---------------------------------------------------------------------------
// Summary
// ---------------------------------------------------------------------------

export function formatRunSummary(run: Run): string {
  const lines = [
    `run ${run.id}: ${run.status}`,
    `  steps_completed: ${String(run.steps_completed)} / ${String(run.max_steps)}`,
    `  credits_charged: ${String(run.credits_charged)} cr`,
    `  cost_cents:      ${String(run.cost_cents)} (${formatUsd(run.cost_cents)})`,
  ];
  if (run.result !== null) {
    lines.push(`  result: passed=${String(run.result.passed)} — ${run.result.summary}`);
    if (run.result.verdict !== undefined && run.result.verdict !== null) {
      lines.push(`  verdict: ${run.result.verdict}`);
    }
  }
  if (run.error !== null) lines.push(`  error: ${run.error.code}: ${run.error.message}`);
  if (run.awaiting_human_reason !== null) {
    lines.push(`  awaiting_human_reason: ${run.awaiting_human_reason}`);
  }
  lines.push(`  request_id: ${run.request_id ?? 'n/a'}`);
  return lines.join('\n');
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

export interface Ex05Config {
  machineId: string;
  task: string;
  cuaVersion: CuaVersion;
  maxSteps: number;
  events: boolean;
  confirm: boolean;
  resumeNote: string;
  pollIntervalMs: number;
}

function intOption(name: string, raw: string | undefined, fallback: number): number {
  if (raw === undefined || raw.trim() === '') return fallback;
  const value = Number(raw);
  if (!Number.isInteger(value) || value < 1) {
    throw new Error(`${name} must be a positive integer, got: ${raw}`);
  }
  return value;
}

export function parseArgs(
  argv: readonly string[],
  env: NodeJS.ProcessEnv = process.env,
): Ex05Config {
  const config: Ex05Config = {
    machineId: env.COASTY_MACHINE_ID ?? '',
    task: env.EX05_TASK ?? 'Open the dashboard and export this month’s report as CSV.',
    cuaVersion: 'v3',
    maxSteps: intOption('EX05_MAX_STEPS', env.EX05_MAX_STEPS, 10),
    events: false,
    confirm: false,
    resumeNote: 'Resolved by the cookbook operator — please continue.',
    pollIntervalMs: 2000,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = (): string => {
      i += 1;
      const value = argv[i];
      if (value === undefined) throw new Error(`missing value for ${String(arg)}`);
      return value;
    };
    switch (arg) {
      case '--machine':
      case '-m':
        config.machineId = next();
        break;
      case '--task':
      case '-t':
        config.task = next();
        break;
      case '--v4':
        config.cuaVersion = 'v4';
        break;
      case '--max-steps':
        config.maxSteps = intOption('--max-steps', next(), config.maxSteps);
        break;
      case '--events':
        config.events = true;
        break;
      case '--confirm':
        config.confirm = true;
        break;
      case '--resume-note':
        config.resumeNote = next();
        break;
      case '--poll-interval-ms':
        config.pollIntervalMs = intOption('--poll-interval-ms', next(), config.pollIntervalMs);
        break;
      default:
        throw new Error(
          `unknown argument: ${String(arg)} (expected --machine, --task, --v4, ` +
            '--max-steps, --events, --confirm, --resume-note, --poll-interval-ms)',
        );
    }
  }
  if (config.machineId === '') {
    throw new Error(
      'a machine id is required: pass --machine <id> or set COASTY_MACHINE_ID. ' +
        'Sandbox keys get an instant free VM from POST /v1/machines (ids look like mch_test_*).',
    );
  }
  return config;
}

export function describeError(error: unknown): string {
  if (error instanceof CoastyError) {
    return `${error.code}: ${error.message} (request_id=${error.requestId ?? 'n/a'})`;
  }
  return error instanceof Error ? error.message : String(error);
}

export async function main(argv: readonly string[] = process.argv.slice(2)): Promise<number> {
  const config = parseArgs(argv);
  const apiKey = getApiKey();
  if (config.cuaVersion === 'v4') stdoutPrint(V4_TIER_NOTE);

  const approved = ensureSpendApproved({
    apiKey,
    confirmFlag: config.confirm,
    items: [
      {
        label:
          `run: <= ${String(config.maxSteps)} steps @ ${config.cuaVersion} ` +
          `(${String(runStepCredits(config.cuaVersion))} cr/step, no other surcharges)`,
        credits: estimateRunCredits({ steps: config.maxSteps, cuaVersion: config.cuaVersion }),
      },
    ],
  });
  if (!approved) return 1;

  const client = new CoastyClient({ apiKey });
  const run = await createRun({
    client,
    request: buildCreateRunRequest({
      machineId: config.machineId,
      task: config.task,
      cuaVersion: config.cuaVersion,
      maxSteps: config.maxSteps,
    }),
  });

  // Demo policy: auto-resume paused runs once the "human" (this CLI) leaves a
  // note. A real operator UI would inspect run.awaiting_human_reason first.
  const onAwaitingHuman: AwaitingHumanHandler = (paused) => {
    stdoutPrint(`human needed on run ${paused.id}: ${paused.awaiting_human_reason ?? 'unknown'}`);
    return Promise.resolve({ note: config.resumeNote });
  };

  let finalRun: Run;
  if (config.events) {
    const watched = await watchRunEvents({ client, runId: run.id, onAwaitingHuman });
    stdoutPrint(`observed ${String(watched.seqs.length)} event(s); billing events:`);
    for (const billing of watched.billingEvents) stdoutPrint(`  ${JSON.stringify(billing)}`);
    finalRun = watched.run;
  } else {
    finalRun = await pollRunUntilTerminal({
      client,
      runId: run.id,
      intervalMs: config.pollIntervalMs,
      onAwaitingHuman,
    });
  }

  stdoutPrint(formatRunSummary(finalRun));
  return finalRun.status === 'succeeded' ? 0 : 1;
}

/** True when this module is the CLI entrypoint (vs being imported by tests). */
function isCliEntry(moduleUrl: string): boolean {
  const entry = process.argv[1];
  if (entry === undefined) return false;
  return pathToFileURL(path.resolve(entry)).href.toLowerCase() === moduleUrl.toLowerCase();
}

if (isCliEntry(import.meta.url)) {
  void main()
    .then((code) => {
      process.exitCode = code;
    })
    .catch((error: unknown) => {
      console.error(describeError(error));
      process.exitCode = 1;
    });
}
